src/core/sock.c

```diff
@@ -1088,3 +1161,24 @@ void nn_sock_stat_increment (struct nn_sock *self, int name, int64_t increment)
             break;
     }
 }
+
+int nn_sock_hold (struct nn_sock *self)
+{
+    switch (self->state) {
+    case NN_SOCK_STATE_ACTIVE:
+    case NN_SOCK_STATE_INIT:
+        self->holds++;
+        return 0;
+    case NN_SOCK_STATE_ZOMBIE:
+        return -ETERM;
+    }
```

Comment:
Agree with all you say, and for clarification, I still delineate the important difference you mention between panic and logging.

The point I was trying to make is this -- on usage of `nn_assert()` currently in libnanomsg in fairly benign failure modes -- let's say a small, infrequent memory leak -- those instances would be changed to a new `nn_log_failure()` or such.

For true sanity checks, where insanity presents a grievous failure mode, `nn_assert()` keeps the exact same behavior as today, and remains in the codebase.

Ideally, this is all moot, with the number of problems resolved in recent weeks!